Intermittent failure of AhcWsProducerConsumer

Pops up now and again on CI builds:

```
2020-09-22T03:22:10.4678375Z [ERROR] testAhcWsProducerConsumer  Time elapsed: 5.067 s  <<< FAILURE!
2020-09-22T03:22:10.4684715Z java.lang.AssertionError: 
2020-09-22T03:22:10.4689652Z 1 expectation failed.
2020-09-22T03:22:10.4698816Z Response body doesn't match expectation.
2020-09-22T03:22:10.4711279Z Expected: is "Hello Camel Quarkus AHC-WS"
2020-09-22T03:22:10.4716236Z   Actual: 
2020-09-22T03:22:10.4718262Z 
2020-09-22T03:22:10.4736164Z 	at org.apache.camel.quarkus.component.http.it.HttpTest.testAhcWsProducerConsumer(HttpTest.java:89)
2020-09-22T03:22:10.4740814Z 
2020-09-22T03:22:10.4774545Z 2020-09-22 03:22:10,476 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (main) Apache Camel 3.5.0 (camel-1) is shutting down
2020-09-22T03:22:10.4856268Z 2020-09-22 03:22:10,484 INFO  [org.apa.cam.com.net.SingleTCPNettyServerBootstrapFactory] (main) ServerBootstrap unbinding from 0.0.0.0:44697
2020-09-22T03:22:10.4917803Z 2020-09-22 03:22:10,488 INFO  [org.apa.cam.com.net.NettyConsumer] (main) Netty consumer unbound from: 0.0.0.0:44697
2020-09-22T03:22:10.4934183Z 2020-09-22 03:22:10,492 INFO  [org.apa.cam.com.net.SingleTCPNettyServerBootstrapFactory] (main) ServerBootstrap unbinding from 0.0.0.0:34351
2020-09-22T03:22:10.5078221Z 2020-09-22 03:22:10,504 INFO  [org.apa.cam.com.net.NettyConsumer] (main) Netty consumer unbound from: 0.0.0.0:34351
2020-09-22T03:22:10.5537175Z 2020-09-22 03:22:10,553 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (main) Apache Camel 3.5.0 (camel-1) uptime 9.538 seconds
2020-09-22T03:22:10.5541564Z 2020-09-22 03:22:10,553 INFO  [org.apa.cam.imp.eng.AbstractCamelContext] (main) Apache Camel 3.5.0 (camel-1) is shutdown in 0.077 seconds
2020-09-22T03:22:10.5691479Z 2020-09-22 03:22:10,568 INFO  [io.quarkus] (main) Quarkus stopped in 0.108s
2020-09-22T03:22:10.6691602Z [INFO] 
2020-09-22T03:22:10.6693206Z [INFO] Results:
2020-09-22T03:22:10.6694033Z [INFO] 
2020-09-22T03:22:10.6695018Z [ERROR] Failures: 
2020-09-22T03:22:10.6696726Z [ERROR]   HttpTest.testAhcWsProducerConsumer:89 1 expectation failed.
2020-09-22T03:22:10.6699253Z Response body doesn't match expectation.
2020-09-22T03:22:10.6700781Z Expected: is "Hello Camel Quarkus AHC-WS"
2020-09-22T03:22:10.6703419Z   Actual: 
2020-09-22T03:22:10.6703738Z 
2020-09-22T03:22:10.6704121Z [INFO] 
```

--- a/integration-tests/http/src/main/java/org/apache/camel/quarkus/component/http/it/GreetingServerEndpoint.java
+++ b/integration-tests/http/src/main/java/org/apache/camel/quarkus/component/http/it/GreetingServerEndpoint.java
@@ -16,14 +16,34 @@
  */
 package org.apache.camel.quarkus.component.http.it;
 
+import javax.websocket.OnClose;
 import javax.websocket.OnMessage;
+import javax.websocket.OnOpen;
+import javax.websocket.Session;
 import javax.websocket.server.ServerEndpoint;
+
+import org.jboss.logging.Logger;
 
 @ServerEndpoint("/ahc-ws/greeting")
 public class GreetingServerEndpoint {
 
+    public static volatile boolean connected = false;
+    private static final Logger LOG = Logger.getLogger(GreetingServerEndpoint.class);
+
+    @OnOpen
+    public void onOpen(Session session) {
+        LOG.infof("WebSocket connection opened for session %s", session.getId());
+        connected = true;
+    }
+
+    @OnClose
+    public void onClose(Session session) {
+        LOG.infof("WebSocket connection closed for session %s", session.getId());
+        connected = false;
+    }
+
     @OnMessage
-    public String onMessage(String message) {
-        return "Hello " + message;
+    public void onMessage(Session session, String message) {
+        session.getAsyncRemote().sendText("Hello " + message);
     }
 }

--- a/integration-tests/http/src/main/java/org/apache/camel/quarkus/component/http/it/HttpResource.java
+++ b/integration-tests/http/src/main/java/org/apache/camel/quarkus/component/http/it/HttpResource.java
@@ -16,6 +16,9 @@
  */
 package org.apache.camel.quarkus.component.http.it;
 
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
 import javax.enterprise.context.ApplicationScoped;
 import javax.inject.Inject;
 import javax.ws.rs.Consumes;
@@ -88,13 +91,24 @@ public class HttpResource {
     @POST
     @Produces(MediaType.TEXT_PLAIN)
     public String triggerAhcWsProducerConsumer(@QueryParam("test-port") int port, String message) throws Exception {
-        producerTemplate
-                .to("direct:ahcWsIn")
-                .withBody(message)
-                .withHeader("test-port", port)
-                .send();
+        String uri = String.format("ahc-ws:localhost:%d/ahc-ws/greeting", port);
 
-        return consumerTemplate.receiveBody("ahc-ws:localhost:" + port + "/ahc-ws/greeting", 5000, String.class);
+        // Start consumer
+        CompletableFuture<String> future = CompletableFuture
+                .supplyAsync(() -> consumerTemplate.receiveBody(uri, 10000, String.class));
+
+        // Wait for consumer connect
+        int attempts = 0;
+        while (!GreetingServerEndpoint.connected && attempts < 25) {
+            Thread.sleep(250);
+            attempts++;
+        }
+
+        // Send WS payload
+        producerTemplate.to(uri).withBody(message).send();
+
+        // Get result from the consumer
+        return future.get(5, TimeUnit.SECONDS);
     }
 
     // *****************************

--- a/integration-tests/http/src/main/java/org/apache/camel/quarkus/component/http/it/HttpRoute.java
+++ b/integration-tests/http/src/main/java/org/apache/camel/quarkus/component/http/it/HttpRoute.java
@@ -43,9 +43,6 @@ public class HttpRoute extends RouteBuilder {
                         exchange.getMessage().setBody(json);
                     }
                 });
-
-        from("direct:ahcWsIn")
-                .toD("ahc-ws:localhost:${header.test-port}/ahc-ws/greeting");
     }
 
     @Named

